src/util/finalizable_processor.rs

```diff
@@ -87,20 +78,50 @@ impl FinalizableProcessor {
     pub fn forward_candidate<E: ProcessEdgesWork>(&mut self, e: &mut E, _nursery: bool) {
         self.candidates
             .iter_mut()
-            .for_each(|reff| *reff = FinalizableProcessor::get_forwarded_finalizable(e, *reff));
+            .for_each(|f| FinalizableProcessor::<F>::forward_finalizable_reference(e, f));
         e.flush();
     }
 
     pub fn forward_finalizable<E: ProcessEdgesWork>(&mut self, e: &mut E, _nursery: bool) {
         self.ready_for_finalize
             .iter_mut()
-            .for_each(|reff| *reff = FinalizableProcessor::get_forwarded_finalizable(e, *reff));
+            .for_each(|f| FinalizableProcessor::<F>::forward_finalizable_reference(e, f));
         e.flush();
     }
 
-    pub fn get_ready_object(&mut self) -> Option<ObjectReference> {
+    pub fn get_ready_object(&mut self) -> Option<F> {
         self.ready_for_finalize.pop()
     }
+
+    pub fn get_all_finalizers(&mut self) -> Vec<F> {
+        let mut ret = std::mem::take(&mut self.candidates);
+        let ready_objects = std::mem::take(&mut self.ready_for_finalize);
+
+        ret.extend(ready_objects);
+        ret
+    }
+
+    pub fn get_finalizers_for(&mut self, object: ObjectReference) -> Vec<F> {
+        // Drain filter for finalizers that equal to 'object':
+        // * for elements that equal to 'object', they will be removed from the original vec, and returned.
+        // * for elements that do not equal to 'object', they will be left in the original vec.
+        let drain_filter = |vec: &mut Vec<F>| -> Vec<F> {
+            let mut i = 0;
+            let mut ret = vec![];
+            while i < vec.len() {
+                if vec[i].get_reference() == object {
+                    let val = vec.remove(i);
```

Comment:
`retain` probably won't work here, as it does not give you the element `T` (you only have `&mut T` in its predicate). I think the only way to get an element from the vec is `remove` or `swap_remove`. `swap_remove` should work fine, but it will unnecessarily shuffle the elements.

Reply:
The pattern I was using is recommended in `vec.drain_filter()`, though it does mention that `drain_filter()` is more efficient: https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.drain_filter

Reply:
OK. If that's recommended, we can keep it. I assume this method is not used very frequently.

Reply:
I will leave it as it is. This method is used to pop finalisers for a given object. For example, in Julia, you can register object `x` with `f()`, `g()`, `h()`. And you can call `finalize(x)`, which should call the registered finalize methods for `x`. The binding can use this method to get all the registered finalizers for `x`. I think it would be better if we preserve the order of the finalizers just in case the runtime/language has some assumption about the order.